graph_weather/models/atmorep/utils/uncertainity.py

```diff
@@ -0,0 +1,467 @@
+import math
+
+import torch
+
+from ..config import AtmoRepConfig
+
+
+class UncertaintyEstimator:
+    def __init__(self, config: AtmoRepConfig):
+        self.config = config
+
+    def estimate_uncertainty(self, ensemble_predictions):
+        """Estimate uncertainty from ensemble predictions"""
+        uncertainties = {}
+
+        for field, preds in ensemble_predictions.items():
+            # Calculate mean and variance across ensemble dimension
+            mean_pred = preds.mean(dim=0)  # [B, T, H, W]
+            variance = ((preds - mean_pred.unsqueeze(0)) ** 2).mean(dim=0)  # [B, T, H, W]
+
+            # Calculate additional uncertainty metrics
+            ensemble_spread = preds.std(dim=0)  # Standard deviation across ensemble
+            entropy = self._calculate_entropy(preds)  # Information entropy
+
+            uncertainties[field] = {
+                "variance": variance,
+                "spread": ensemble_spread,
+                "entropy": entropy,
+            }
+
+        return uncertainties
+
+    def _calculate_entropy(self, ensemble_preds, num_bins=10):
+        """Calculate entropy by binning predictions and computing Shannon entropy"""
+        # For continuous variables, we discretize into bins
+        E, B, T, H, W = ensemble_preds.shape
+
+        # Reshape for binning
+        flat_preds = ensemble_preds.reshape(E, -1)
+
+        # Find global min and max for consistent binning
+        global_min = flat_preds.min()
+        global_max = flat_preds.max()
+        bin_width = (global_max - global_min) / num_bins
+
+        # Initialize entropy map
+        entropy = torch.zeros(B, T, H, W)
+
+        # Calculate entropy for each pixel location
```

Comment:
This will be incredibly slow with larger input grids, please refactor this to use vectorized operations.